go/pkg/ddapi/translate.go

```diff
@@ -0,0 +1,208 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ddapi
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	json "github.com/json-iterator/go"
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/prometheus/prompb"
+)
+
+type ddPoint struct {
+	Timestamp int64
+	Value     float64
+}
+
+// In the expected JSON document structure, an individual sample/point is
+// provided as a 2-tuple via an array. For example: [1610032230, 0.1]. One
+// can't just `json.Unmarshal()` an array into a struct. Use a custom
+// deserialization method for `ddPoint`, to strictly check the length of the n
+// tuple (array) and types of its elements. Kudos to
+// https://eagain.net/articles/go-json-array-to-struct/  -- relevant docs: "To
+// unmarshal JSON into a value implementing the Unmarshaler interface,
+// Unmarshal calls that value's UnmarshalJSON method, including when the input
+// is a JSON null".
+func (p *ddPoint) UnmarshalJSON(buf []byte) error {
+	dp := []interface{}{&p.Timestamp, &p.Value}
+
+	if err := json.Unmarshal(buf, &dp); err != nil {
+		return err
+	}
+
+	if len(dp) != 2 {
+		return fmt.Errorf("unexpected length pf `points` array: %d (expected 2)", len(dp))
+	}
+
+	return nil
+}
+
+type ddSeriesFragment struct {
+	Name           string    `json:"metric"`
+	Points         []ddPoint `json:"points"`
+	Tags           []string  `json:"tags"`
+	Host           string    `json:"host"`
+	Device         string    `json:"device,omitempty"`
+	Type           string    `json:"type"`
+	Interval       int64     `json:"interval"`
+	SourceTypeName string    `json:"source_type_name,omitempty"`
+}
+
+type ddSeriesFragments struct {
+	Fragments []*ddSeriesFragment `json:"series"`
+}
+
+var invalidCharRE = regexp.MustCompile(`[^a-zA-Z0-9_:]`)
+
+func sanitizeMetricName(value string) string {
+	return invalidCharRE.ReplaceAllString(value, "_")
+}
+
+func sanitizeLabelName(value string) string {
+	return invalidCharRE.ReplaceAllString(value, "_")
+}
+
+/*
+
+Expected input JSON document structure: a set of time series fragments
+that the DD agent POSTs to /api/v1/series. Example:
+
+{
+  "series": [
+    {
+      "metric": "datadog.trace_agent.trace_writer.traces",
+      "points": [
+        [
+          1610032230,
+          0
+        ],
+        ...
+      ],
+      "tags": [
+        "version:7.24.1"
+      ],
+      "host": "x1carb6",
+      "type": "rate",
+      "interval": 10
+    },
+    {
+      "metric": "datadog.dogstatsd.client.metrics",
+      "points": [
+    ...
+    }
+  ]
+}
+
+Note that each object in the `series` array is a time series fragment for a
+specific metric. Each metric is defined by its name and other meta data (think:
+labels). The time series fragment is comprised of one or multiple data points /
+samples.
+*/
+func TranslateDDSeriesJSON(doc []byte) ([]*prompb.TimeSeries, error) {
```

Comment:
I'll suggest starting to write tests for this function as soon as possible because it's on the hot path. Golang's test harness also makes it easier to [write benchmarks to understand this function memory allocation profile](https://golang.org/pkg/testing/#AllocsPerRun).

Reply:
Yes! :fire: